README.md
Copyright (c) 2013 Roland van Rijswijk-Deij

All rights reserved. This software is distributed under a BSD-style
license. For more information, see LICENSE

1. INTRODUCTION
===============

EDNA is the Emulator Daemon for NFC Applications. It emulates an NFC smart card
using the SpringCard NFC'Roll reader device.

2. PREREQUISITES
================

To build the library:

 - PC/SC lite (>= 1.8.3)
 - libccid (>= 1.4.10) (not a direct dependency; required for the SpringCard NFC'Roll)

3. BUILDING
===========

To build the library, execute the following commands:

    ./autogen.sh
    ./configure
    make

4. INSTALLING
=============

To install the library as a regular user, run:

    sudo make install

If you are root (administrative user), run:

    make install

Finally, you have to configure the generic CCID smart card reader driver to allow
control commands. Go to the reader driver package directory (usually /usr/lib/pcsc/drivers/ifd-ccid.bundle), and
go into the Contents subdirectory. Open the file called Info.plist in an editor and look for the following lines:

    <key>ifdDriverOptions</key>
    <string>0x0000</string>

Changes this to:

    <key>ifdDriverOptions</key>
    <string>0x0001</string>

You may have to restart PC/SC lite after this change.

A sample configuration file is installed in the <prefix>/share/doc/silvia directory (usually /usr/share/doc/silvia);
please refer to this file for information on how to configure your edna installation.

5. USING THE APPLICATION
========================

T.B.D.

6. CONTACT
==========

Questions/remarks/suggestions/praise on this tool can be sent to:

Roland van Rijswijk-Deij <rijswijk@cs.ru.nl, roland@mazuki.nl>


ChangeLog
2013-09-08	Roland van Rijswijk-Deij <rijswijk@cs.ru.nl>

	* Created initial autotoolized version
	* Skeleton


include/edna.h
/*
 * Copyright (c) 2013 Roland van Rijswijk-Deij
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

/*
 * The Emulator Daemon for NFC Applications (EDNA)
 * General include file
 */

#ifndef _EDNA_H
#define _EDNA_H

#include "config.h"
#include <stdlib.h>

#define FLAG_SET(flags, flag) ((flags & flag) == flag)

/* Type for function return values */
typedef unsigned long edna_rv;

/* Exported functions */

#ifdef __cplusplus
extern "C" 
{
#endif // __cplusplus

/**
 * Initialise the edna client library
 * @return ERV_OK on success, an appropriate error otherwise
 */
edna_rv edna_lib_init(void);

/**
 * Uninitialise the edna client library
 * @return ERV_OK on success, an appropriate error otherwise
 */
edna_rv edna_lib_uninit(void);

/* Types for callback functions */
typedef int (*handle_apdu)(const unsigned char* apdu_data, size_t apdu_len, unsigned char* rdata, size_t* rdata_len);

/**
 * Connect to the daemon and register an AID
 * @param aid_data the AID data
 * @param aid_len the length of the AID data
 * @return ERV_OK if the connection was established, an appropriate error otherwise
 */
edna_rv edna_lib_connect(const unsigned char* aid_data, size_t aid_len);

/**
 * Disconnect from the daemon (unregisters the previously registered AID)
 * @return ERV_OK if disconnect was successful, an appropriate error otherwise
 */
edna_rv edna_lib_disconnect(void);

/**
 * Process APDUs sent by the daemon; this function runs an event loop
 * that calls the supplied callback function to process the APDUs.
 * The loop can be terminated by a call to edna_lib_cancel.
 * @param process_cb the callback function that processes APDUs
 * @return ERV_OK if the event loop terminated because of a call to
 *         edna_lib_cancel, an error indicating why the event loop
 *         was terminated otherwise
 */
edna_rv edna_lib_loop_and_process(handle_apdu process_cb);

/**
 * Terminate the event loop
 */
void edna_lib_cancel(void);

#ifdef __cplusplus
}
#endif // __cplusplus

/* Function return values */

/* Success */
#define ERV_OK					0x00000000

/* Warning messages */

#define ERV_ALREADY_INITIALISED	0x40000000	/* The client library was already initialised */

/* Error messages */

/* General errors */
#define ERV_GENERAL_ERROR		0x80000000	/* An undefined error occurred */
#define ERV_MEMORY				0x80000001	/* An error occurred while allocating memory */
#define ERV_PARAM_INVALID		0x80000002	/* Invalid parameter(s) provided for function call */
#define ERV_LOG_INIT_FAIL		0x80000003	/* Failed to initialise logging */
#define ERV_NOT_INITIALISED		0x80000004	/* The client library is not initialised */

/* Configuration errors */
#define ERV_NO_CONFIG			0x80001000	/* No configuration file was specified */
#define ERV_CONFIG_ERROR		0x80001001	/* An error occurred while reading the configuration file */
#define ERV_CONFIG_NO_ARRAY		0x80001002	/* The requested configuration item is not an array */
#define ERV_CONFIG_NO_STRING	0x80001003	/* The requested configuration item is not a string */

/* Library errors */
#define ERV_CONNECT_FAILED		0x80002000	/* Failed to connect to the daemon */
#define ERV_DISCONNECTED		0x80002001	/* The connection with the daemon was closed unexpectedly */
#define ERV_NOT_CONNECTED		0x80002002	/* There is no connection to the daemon */
#define ERV_VERSION_MISMATCH	0x80002003	/* The daemon reported a mismatching API version */
#define ERV_ALREADY_CONNECTED	0x80002004	/* There is already a connection to the daemon */
#define ERV_ALREADY_REGISTERED	0x80002005	/* The specified AID has already been registered with the daemon */

#endif /* !_EDNA_H */



src/common/edna_proto.h
/*
 * Copyright (c) 2013 Roland van Rijswijk-Deij
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

/*
 * The Emulator Daemon for NFC Applications (EDNA)
 * Daemon <-> Library protocol 
 */

#ifndef _EDNA_PROTO_H
#define _EDNA_PROTO_H

/* UNIX domain socket name */
#define EDNA_SOCKET			"/tmp/edna-comm"

#ifndef UNIX_PATH_MAX
#define UNIX_PATH_MAX 		80 			/* should be safe */
#endif // !UNIX_PATH_MAX

/* API version */
#define API_VERSION			0x00

/* API commands */
#define GET_API_VERSION		0x01
#define REGISTER_AID		0x02

/* API return values */
#define EDNA_OK				0x00
#define	AID_EXISTS			0x01

#endif // !_EDNA_PROTO_H



src/common/edna_bytestring.h
/*
 * Copyright (c) 2013 Roland van Rijswijk-Deij
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

/*
 * The Emulator Daemon for NFC Applications (EDNA)
 * Byte string class
 */

#ifndef _EDNA_BYTESTRING_H
#define _EDNA_BYTESTRING_H

#include <cstddef>
#include <vector>
#include <string>
#include <stdlib.h>
#include <limits.h>
#include <gmpxx.h>
#include "config.h"

#ifndef SIZE_T_MAX
#define SIZE_T_MAX ((size_t) -1)
#endif // !SIZE_T_MAX

class bytestring
{
public:
	// Constructors
	bytestring();

	bytestring(const unsigned char* bytes, const size_t bytesLen);

	bytestring(const char* hexString);

	bytestring(const unsigned long longValue);

	bytestring(const bytestring& in);
	
	// Destructor
	virtual ~bytestring() { }

	// Append data
	bytestring& operator+=(const bytestring& append);
	bytestring& operator+=(const unsigned char byte);
	bytestring& operator+=(const unsigned short ushort_val);

	// Return a substring
	bytestring substr(const size_t start, const size_t len = SIZE_T_MAX) const;

	// Array operator
	unsigned char& operator[](size_t pos);

	// Return the byte string
	unsigned char* byte_str();
	
	// Return the const byte string
	const unsigned char* const_byte_str() const;

	// Return a hexadecimal character representation of the string
	std::string hex_str() const;
	
	// Split of the specified part of the string as a separate byte string
	bytestring split(size_t len);

	// Return the size in bits
	size_t bits() const;

	// Return the size in bytes
	size_t size() const;

	// Resize
	void resize(const size_t newSize);

	// Wipe
	void wipe(const size_t newSize = 0);

	// Comparison
	bool operator==(const bytestring& compareTo) const;
	bool operator!=(const bytestring& compareTo) const;
	bool operator<(const bytestring& compareTo) const;
	bool operator>(const bytestring& compareTo) const;

	// XORing
	bytestring& operator^=(const bytestring& rhs);

private:
	std::vector<unsigned char> byteString;
};

// Add data
bytestring operator+(const bytestring& lhs, const bytestring& rhs);
bytestring operator+(const unsigned char lhs, const bytestring& rhs);
bytestring operator+(const bytestring& lhs, const unsigned char rhs);

// XOR data
bytestring operator^(const bytestring& lhs, const bytestring& rhs);

#endif // !_EDNA_BYTESTRING_H



src/common/edna_bytestring.cpp
/*
 * Copyright (c) 2013 Roland van Rijswijk-Deij
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

/*
 * The Emulator Daemon for NFC Applications (EDNA)
 * Byte string class
 */

#include <algorithm>
#include <string>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "config.h"
#include "edna_bytestring.h"

// Constructors
bytestring::bytestring()
{
}

bytestring::bytestring(const unsigned char* bytes, const size_t bytesLen)
{
	byteString.resize(bytesLen);

	memcpy(&byteString[0], bytes, bytesLen);
}

bytestring::bytestring(const char* hexString)
{
	std::string hex = std::string(hexString);

	if (hex.size() % 2 != 0)
	{
		hex = "0" + hex;
	}

	for (size_t i = 0; i < hex.size(); i += 2)
	{
		std::string byteStr;
		byteStr += hex[i];
		byteStr += hex[i+1];

		unsigned char byteVal = (unsigned char) strtoul(byteStr.c_str(), NULL, 16);

		this->operator+=(byteVal);
	}
}

bytestring::bytestring(const unsigned long longValue)
{
	unsigned long setValue = longValue;

	// Convert the value to a big-endian byte string; N.B.: this code assumes that unsigned long
	// values are stored as a 64-bit value, which is a safe assumption on modern systems. It will
	// also properly handle a 32-bit value and will simply store 4 zeroes at the front of the
	// string. If at some point in time we get 128-bit architectures, the top 8 bytes of the value
	// will be discarded... (but hey, 640K is enough for everybody, right?)
	//
	// The reason for coding it this way is that implementations of SoftHSM will maintain
	// binary compatibility between eachothers background storage (i.e. a 32-bit SoftHSM version can
	// read the storage of a 64-bit version and vice versa under the assumption that the stored
	// values never exceed 32-bits, which is likely since these values are only used to encode
	// byte string lengths)
	unsigned char byteStrIn[8];
	
	for (size_t i = 0; i < 8; i++)
	{
		byteStrIn[7-i] = (unsigned char) (setValue & 0xFF);
		setValue >>= 8;
	}

	byteString.resize(8);
	memcpy(&byteString[0], byteStrIn, 8);
}

bytestring::bytestring(const bytestring& in)
{
	this->byteString = in.byteString;
}

// Append data
bytestring& bytestring::operator+=(const bytestring& append)
{
	size_t curLen = byteString.size();
	size_t toAdd = append.byteString.size();
	size_t newLen = curLen + toAdd;

	byteString.resize(newLen);

	memcpy(&byteString[curLen], &append.byteString[0], toAdd);

	return *this;
}

bytestring& bytestring::operator+=(const unsigned char byte)
{
	byteString.push_back(byte);

	return *this;
}

bytestring& bytestring::operator+=(const unsigned short ushort_val)
{
	this->operator+=((unsigned char) ((ushort_val >> 8) & 0xff));
	this->operator+=((unsigned char) (ushort_val & 0xff));
}

// XORing
bytestring& bytestring::operator^=(const bytestring& rhs)
{
	size_t xorLen = std::min(this->size(), rhs.size());

	for (size_t i = 0; i < xorLen; i++)
	{
		byteString[i] ^= rhs.const_byte_str()[i];
	}

	return *this;
}

// Return a substring
bytestring bytestring::substr(const size_t start, const size_t len /* = SIZE_T_MAX */) const
{
	size_t retLen = std::min(len, byteString.size() - start);

	if (start >= byteString.size())
	{
		return bytestring();
	}
	else
	{
		return bytestring(&byteString[start], retLen);
	}
}

// Add data
bytestring operator+(const bytestring& lhs, const bytestring& rhs)
{
	bytestring rv = lhs;
	rv += rhs;

	return rv;
}

bytestring operator+(const unsigned char lhs, const bytestring& rhs)
{
	bytestring rv(&lhs, 1);
	rv += rhs;

	return rv;
}

bytestring operator+(const bytestring& lhs, const unsigned char rhs)
{
	bytestring rv = lhs;
	rv += rhs;

	return rv;
}

// Array operator
unsigned char& bytestring::operator[](size_t pos)
{
	return byteString[pos];
}

// Return the byte string data
unsigned char* bytestring::byte_str()
{
	return &byteString[0];
}

// Return the const byte string
const unsigned char* bytestring::const_byte_str() const
{
	return (const unsigned char*) &byteString[0];
}

// Return a hexadecimal character representation of the string
std::string bytestring::hex_str() const
{
	std::string rv;
	char hex[3];

	for (size_t i = 0; i < byteString.size(); i++)
	{
		sprintf(hex, "%02X", byteString[i]);

		rv += hex;
	}

	return rv;
}

// Split of the specified part of the string as a separate byte string
bytestring bytestring::split(size_t len)
{
	bytestring rv = substr(0, len);

	size_t newSize = (byteString.size() > len) ? (byteString.size() - len) : 0;

	if (newSize > 0)
	{
		for (size_t i = 0; i < newSize; i++)
		{
			byteString[i] = byteString[i + len];
		}
	}

	byteString.resize(newSize);

	return rv;
}

// The size of the byte string in bits
size_t bytestring::bits() const
{
	size_t bits = byteString.size() * 8;

	if (bits == 0) return 0;

	for (size_t i = 0; i < byteString.size(); i++)
	{
		unsigned char byte = byteString[i];

		for (unsigned char mask = 0x80; mask > 0; mask >>= 1)
		{
			if ((byte & mask) == 0)
			{
				bits--;
			}
			else
			{
				return bits;
			}
		}
	}

	return bits;
}

// The size of the byte string in bytes
size_t bytestring::size() const
{
	return byteString.size();
}

void bytestring::resize(const size_t newSize)
{
	byteString.resize(newSize);
}

void bytestring::wipe(const size_t newSize /* = 0 */)
{
	this->resize(newSize);

	memset(&byteString[0], 0x00, byteString.size());
}

// Comparison
bool bytestring::operator==(const bytestring& compareTo) const
{
	if (compareTo.size() != this->size())
	{
		return false;
	}

	return (memcmp(&byteString[0], &compareTo.byteString[0], this->size()) == 0);
}

bool bytestring::operator!=(const bytestring& compareTo) const
{
	if (compareTo.size() != this->size())
	{
		return true;
	}

	return (memcmp(&byteString[0], &compareTo.byteString[0], this->size()) != 0);
}

bool bytestring::operator<(const bytestring& compareTo) const
{
	return this->hex_str() < compareTo.hex_str();
}

bool bytestring::operator>(const bytestring& compareTo) const
{
	return this->hex_str() > compareTo.hex_str();
}

// XOR data
bytestring operator^(const bytestring& lhs, const bytestring& rhs)
{
	size_t xorLen = std::min(lhs.size(), rhs.size());
	bytestring rv;

	for (size_t i = 0; i < xorLen; i++)
	{
		rv += (unsigned char) (lhs.const_byte_str()[i] ^ rhs.const_byte_str()[i]);
	}

	return rv;
}


src/lib/edna_lib_export.cpp
/*
 * Copyright (c) 2013 Roland van Rijswijk-Deij
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

/*
 * The Emulator Daemon for NFC Applications (EDNA)
 * Exported functions
 */

#include "config.h"
#include "edna.h"
#include "edna_proto.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <vector>
#include <errno.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/types.h>


/* Library status */
static bool edna_lib_initialised = false;

static bool edna_lib_connected = false;

static bool edna_lib_must_cancel = false;

/* Connection to the daemon */
int daemon_socket = -1;

edna_rv edna_lib_init(void)
{
	if (edna_lib_initialised)
	{
		return ERV_ALREADY_INITIALISED;
	}
	
	edna_lib_initialised = true;
	edna_lib_connected = false;
	daemon_socket = -1;
	
	return ERV_OK;
}

edna_rv edna_lib_uninit(void)
{
	if (!edna_lib_initialised)
	{
		return ERV_NOT_INITIALISED;
	}
	
	if ((daemon_socket >= 0) && edna_lib_connected)
	{
		close(daemon_socket);
		edna_lib_connected = false;
	}
	
	edna_lib_initialised = false;
	edna_lib_connected = false;
	daemon_socket = -1;
	
	return ERV_OK;
}

int send_to_daemon(const std::vector<unsigned char> tx)
{
	if (!edna_lib_connected || (daemon_socket < 0))
	{
		return -1;
	}
	
	if (tx.size() > 0xffff) return -1;
	
	/* 
	 * Prepare the data for transmission by prepending a 16-bit
	 * value indicating the command length 
	 */
	unsigned short tx_size = (unsigned short) tx.size();
	
	std::vector<unsigned char> tx_buf;
	tx_buf.resize(tx.size() + 2);
	
	tx_buf[0] = tx_size >> 8;
	tx_buf[1] = tx_size & 0xff;
	
	memcpy(&tx_buf[2], &tx[0], tx.size());
	
	/* 
	 * Transmit the command
	 * 
	 * FIXME: this will fail if the process receives a POSIX signal! 
	 */
	int tx_sent = 0;
	 
	if ((tx_sent = write(daemon_socket, &tx_buf[0], tx_buf.size())) != tx_buf.size())
	{
		close(daemon_socket);
		
		edna_lib_connected = false;
		daemon_socket = -1;
		
		return -2;
	}
	
	return 0;
}

int recv_from_daemon(std::vector<unsigned char>& rx)
{
	if (!edna_lib_connected || (daemon_socket < 0))
	{
		return -1;
	}
	
	unsigned short rx_size = 0;
	
	size_t rx_index = 0;
	
	unsigned char buf[512] = { 0 };
	
	/*
	 * Read the length of the data to receive
	 * 
	 * FIXME: this will fail if the process receives a signal!
	 */
	if ((read(daemon_socket, &buf[0], 1) != 1) ||
	    (read(daemon_socket, &buf[1], 1) != 1))
	{
		close(daemon_socket);
		
		edna_lib_connected = false;
		daemon_socket = -1;
		
		return -2;
	}
	
	rx_size = (buf[0] << 8) + buf[1];
	
	rx.resize(rx_size);
	
	/* 
	 * Now receive the actual data 
	 * 
	 * FIXME: again, will not deal well with POSIX signals!
	 */
	while (rx_size > 0)
	{
		int received = read(daemon_socket, &buf[0], (rx_size < 512) ? rx_size : 512);
		
		if (received < 0)
		{
			close(daemon_socket);
		
			edna_lib_connected = false;
			daemon_socket = -1;
			
			return -2;
		}
		
		memcpy(&rx[rx_index], buf, received);
		
		rx_index += received;
		rx_size -= received;
	}
	
	return 0;
}

edna_rv edna_lib_connect(const unsigned char* aid_data, size_t aid_len)
{
	if (edna_lib_connected)
	{
		return ERV_ALREADY_CONNECTED;
	}
	
	/* Attempt to connect to the daemon */
	struct sockaddr_un addr = { 0 };
	
	daemon_socket = socket(PF_UNIX, SOCK_STREAM, 0);
	
	if (daemon_socket < 0)
	{
		daemon_socket = -1;
		
		return ERV_CONNECT_FAILED;
	}
	
	addr.sun_family = AF_UNIX;
	snprintf(addr.sun_path, UNIX_PATH_MAX, EDNA_SOCKET);
	
	if (connect(daemon_socket, (struct sockaddr*) &addr, sizeof(struct sockaddr_un)) != 0)
	{
		close(daemon_socket);
		
		daemon_socket = -1;
		
		return ERV_CONNECT_FAILED;
	}
	
	edna_lib_connected = true;
	
	/* Request the API version from the daemon */
	std::vector<unsigned char> get_api_version;
	get_api_version.push_back(GET_API_VERSION);
	
	if (send_to_daemon(get_api_version) != 0)
	{
		close(daemon_socket);
		
		daemon_socket = -1;
		edna_lib_connected = false;
		
		return ERV_DISCONNECTED;
	}
	
	std::vector<unsigned char> api_version_info;
	
	if (recv_from_daemon(api_version_info) != 0)
	{
		close(daemon_socket);
		
		daemon_socket = -1;
		edna_lib_connected = false;
		
		return ERV_DISCONNECTED;
	}
	
	if ((api_version_info.size() != 1) || (api_version_info[0] != API_VERSION))
	{
		close(daemon_socket);
		
		daemon_socket = -1;
		edna_lib_connected = false;
		
		return ERV_VERSION_MISMATCH;
	}
	
	/* The API version checks out, register the AID with the daemon */
	std::vector<unsigned char> register_aid;
	register_aid.resize(aid_len + 1);
	register_aid[0] = REGISTER_AID;
	memcpy(&register_aid[1], aid_data, aid_len);
	
	if (send_to_daemon(register_aid) != 0)
	{
		close(daemon_socket);
		
		daemon_socket = -1;
		edna_lib_connected = false;
		
		return ERV_DISCONNECTED;
	}
	
	std::vector<unsigned char> register_aid_rv;
	
	if (recv_from_daemon(register_aid_rv) != 0)
	{
		close(daemon_socket);
		
		daemon_socket = -1;
		edna_lib_connected = false;
		
		return ERV_DISCONNECTED;
	}
	
	if ((register_aid_rv.size() != 1) || (register_aid_rv[0] != EDNA_OK))
	{
		close(daemon_socket);
		
		daemon_socket = -1;
		edna_lib_connected = false;
		
		return ERV_ALREADY_REGISTERED;
	}
	
	return ERV_OK;
}

edna_rv edna_lib_disconnect(void)
{
	if (!edna_lib_connected)
	{
		return ERV_NOT_CONNECTED;
	}
	
	close(daemon_socket);
	edna_lib_connected = false;
	
	return ERV_OK;
}

edna_rv edna_lib_loop_and_process(handle_apdu process_cb)
{
	edna_lib_must_cancel = false;
	
	while (!edna_lib_must_cancel)
	{
		fd_set daemon_fds;
	
		do
		{
			FD_ZERO(&daemon_fds);
			FD_SET(daemon_socket, &daemon_fds);
	
			struct timeval timeout = { 0, 10000 }; // 10ms
			
			select(FD_SETSIZE, &daemon_fds, NULL, NULL, &timeout);
		}
		while (!FD_ISSET(daemon_socket, &daemon_fds) && !edna_lib_must_cancel);
		
		if (edna_lib_must_cancel) break;
		
		/* Receive APDU data from the daemon */
		std::vector<unsigned char> apdu_data;
		
		if (recv_from_daemon(apdu_data) != 0)
		{
			close(daemon_socket);
			
			daemon_socket = -1;
			edna_lib_connected = false;
			
			return ERV_DISCONNECTED;
		}
		
		std::vector<unsigned char> rdata;
		rdata.resize(512);
		size_t rdata_len = 512;
		
		(process_cb)(&apdu_data[0], apdu_data.size(), &rdata[0], &rdata_len);
		
		/* Transmit the response data to the daemon */
		rdata.resize(rdata_len);
		
		if (send_to_daemon(rdata) != 0)
		{
			close(daemon_socket);
			
			daemon_socket = -1;
			edna_lib_connected = false;
			
			return ERV_DISCONNECTED;
		}
	}
	
	edna_lib_must_cancel = false;
	
	return ERV_OK;
}

void edna_lib_cancel(void)
{
	edna_lib_must_cancel = true;
}
